### ISSUE TITLE
Edit multiple documents in parallel

### ISSUE THREAD

**Issue by JaCoB1123** [eikek/docspell#253]
It would be nice to be able to edit multiple documents in parallel (e.g. set organization & person or edit tags).

**Comment by eikek** [comment]
Yes, very true. This is a nice addition! It is on my list, but I admit not on the top. It is especially hard (for me) to create a good user interface for this (selecting multiple items).

**Comment by JaCoB1123** [comment]
I'd like to give it a try myself, I don't know where to start though as I'm new to both Scala and Elm. Could you point me to the right files/folders to take a look at?

**Comment by eikek** [comment]
Sure, thanks for trying! Here are some pointers:

- The webapp is in `modules/webapp` – all the `.elm` files in there make up the web-ui
- If you have used things like react, redux, immutablejs [you will love elm](https://ohanhi.com/react-typescript-vs-elm.html) :)
- The [elm guide](https://guide.elm-lang.org/) is a good starting point
- For developing you need to install elm and [sbt](https://scala-sbt.org) and then start sbt in the root of the project
- In sbt shell, type `make` to build everything (this is only needed once)
- Then type `project restserver` to change into the rest-server module and type `reStart` – this will start the rest server only (no need to waste resource with the job executor if it's not needed)
- If the console gets quiter, type `~ compile` – the `~` tells sbt to recompile on change.If you now change an elm file and save it, it gets recompiled automatically. Then you only need to refresh the browser to see the effect.
- if you create a file in `local/dev.conf` this will be used (if you want to change something during development)

Hope that helps for a first start. You can also write me an email if you like.

**Comment by totti4ever** [comment]
Damn, is it that hard to add a checkbox to the overview and then apply the metadata to all selected items? :-/

That would really help A LOT when you are new to Docspell as you have to organize the whole history of files you are bringing with you. I guess this is true for most people, but you. Isn't it, because you built up the history in Docspell itself?

UI looks so easy when people do it who are used to it :-(

**Comment by eikek** [comment]
I imported all my docs into docspell a year ago and then went through them one by one :). But I use tags quite sparingly, maybe this is the reason I felt ok with that, since most correspondents etc were recognized. Lucky me :) But you are so right: new users coming with a bunch of documents have a hard time getting their initial data into docspell.

I also would like to have this, but I really find it hard to do. I think you would want to add items to a kind-of "shopping cart" and then present a detail form that would apply to all items currently in the cart? So you can select items from different searches. I think I would add server-side support for all those "set routes" for a list of item-ids. Then the client only needs to call different endpoints for the same form.

I'd also like to link items to one another, but this is also something I'm not sure how to do ui wise :-) It "feels" like a similar problem.

When coming from other systems, it might be possible to have an importer? If paperless (for example) can output your metadata somehow, it might be possible to write an importer for this?

**Comment by totti4ever** [comment]
> I imported all my docs into docspell a year ago and then went through them one by one :). But I use tags quite sparingly, maybe this is the reason I felt ok with that, since most correspondents etc were recognized. Lucky me :) 

How did you manage to do that? I am not talking about tags. Main priority are correspondents for me as well. So how did you do that: Importing a bunch of files and still having them recognized by AI?

This is actually what holds me from starting using Docspell. If I had the correspondents, tags might be nice to have, but without correspondents right now it is a big mess. And going through them 1 by 1 doesn't sound appealing :-)

**Comment by eikek** [comment]
well I added most of them to the address book first. then scanned some 20-30 docs and went though them fixing missing correspondents. then went on with the next batch.. it wasn't a lot of fun, but was ok for me.

**Comment by totti4ever** [comment]
> I imported all my docs into docspell a year ago and then went through them one by one :). But I use tags quite sparingly, maybe this is the reason I felt ok with that, since most correspondents etc were recognized. Lucky me :) But you are so right: new users coming with a bunch of documents have a hard time getting their initial data into docspell.

Okay, after starting to import proper documents into my system, I have to say that multi-select-edit is a MUST! Reason is quite simple: Everything which you want to add to documents, which has not been there during import or has been there but couldn't be assigned by AI is typically a mass-edit.

For instance, I'd like to use the equipment feature, but only introduce it after all the files have been imported. It would be quite quick, if I just could filter on a couple of things (organisation, tag or whatsoever), then `select all` and apply an equipment to them.
Not having this function would mean to go through all of them with the arrows or build a script... :-(


> I also would like to have this, but I really find it hard to do. I think you would want to add items to a kind-of "shopping cart" and then present a detail form that would apply to all items currently in the cart? So you can select items from different searches. I think I would add server-side support for all those "set routes" for a list of item-ids. Then the client only needs to call different endpoints for the same form.

I would suggest to not make it too complex for a start:
Isn't all we need

1. Some button to switch to `mass edit mode``
2. Then on all items a little unchecked checkbox would appear (while the items stay the same, so if filters have been applied they remain applied)
3. The left sidebar with the filters is replaced with the metadata edit sidebar from the item detail page
4. Some `select all` and `exit mass edit mode` button

Then, one could easily apply filters as need and either further select the items to change or just mark all and finally do the things you wanna do in the sidebar (incl. deleting them). Process-wise you could just queue all the single commands to use the existing API

> I'd also like to link items to one another, but this is also something I'm not sure how to do ui wise :-) It "feels" like a similar problem.

Having tags and correspondents, from my point of view this is not soo important. But, of course, it would be nice, if I had some letters which belong to one case, which I could link together with not having to add tags for each of those chains. But for me "invisible" tags would be enough, i.e. when filtering I can select to not show them to make finding the proper tags easier


> When coming from other systems, it might be possible to have an importer? If paperless (for example) can output your metadata somehow, it might be possible to write an importer for this?
It's all in the database, so doable - stay tuned for paperless. But STILL: I'd like to benefit from the things Docspell offers MORE than paperless (like equipment) and there cannot be imported!

**Comment by eikek** [comment]
@totti4ever I like this idea. I think this is a good start. It allows to mass-edit only what is visible on one page, but this can be extended later, if needed

@JaCoB1123 Did you already start and put effort into this? I would otherwise work on this in the next weeks, but I don't want to take it away from you :-) I'm happily work on something else (or continue your work)!

**Comment by JaCoB1123** [comment]
@eikek I did put some effort in it, but it's probably a too complex first issue. Feel free to tackle it yourself, I won't have time to work on this in the next weeks anyways. I don't have any useful code to show you either. Thanks for your great work so far, btw. 👍

**Comment by eikek** [comment]
@JaCoB1123 Thanks for your reply and your efforts! Ok I'll try then. If you ever like to try some simpler thing to start with, let me know (can choose between Scala, Elm or both :)).
